core/server/middleware/auth.js

```diff
@@ -0,0 +1,120 @@
+var _           = require('lodash'),
+    passport    = require('passport'),
+    url         = require('url'),
+    errors      = require('../errors'),
+    config      = require('../config'),
+    oauthServer,
+
+    auth;
+
+function cacheOauthServer(server) {
+    oauthServer = server;
+}
+
+function isBearerAutorizationHeader(req) {
+    var parts,
+        scheme,
+        credentials;
+
+    if (req.headers && req.headers.authorization) {
+        parts = req.headers.authorization.split(' ');
+    } else {
+        return false;
+    }
+
+    if (parts.length === 2) {
+        scheme = parts[0];
+        credentials = parts[1];
+        if (/^Bearer$/i.test(scheme)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+auth = {
+
+    // ### Authenticate Client Middleware
+    authenticateClient: function authenticateClient(req, res, next) {
+        // skip client authentication if bearer token is present
+        if (isBearerAutorizationHeader(req)) {
+            return next();
+        }
+
+        if (req.query && req.query.client_id) {
+            req.body.client_id = req.query.client_id;
+        }
+
+        if (req.query && req.query.client_secret) {
+            req.body.client_secret = req.query.client_secret;
+        }
+
+        if (!req.body.client_id || !req.body.client_secret) {
+            return errors.handleAPIError(new errors.UnauthorizedError('Access denied.'), req, res, next);
+        }
+
+        return passport.authenticate(['oauth2-client-password'], {session: false, failWithError: false},
+            function authenticate(err, client) {
+                var origin = null;
+                if (err) {
+                    return next(err); // will generate a 500 error
+                }
+
+                if (req.headers && req.headers.origin) {
+                    origin = url.parse(req.headers.origin).hostname;
+                }
+
+                if (origin && client && client.type === 'ua'
+                    && (_.some(client.trustedDomains, {trusted_domain: origin}) || origin === url.parse(config.url).hostname)) {
```

Comment:
I think it makes sense to add `urlSSL` here as well, and as this list of things-that-are-ok is growing quite long, it may be helpful to split it into its own `isValidOrigin` function and test it independently?